src/bin/bobd.rs

```diff
@@ -32,26 +38,52 @@ impl server::BobApi for BobSrv {
         type PingFuture = future::FutureResult<Response<Null>, tower_grpc::Status>;
 
         fn put(&mut self, req: Request<PutRequest>) -> Self::PutFuture {
-            let mut param = req.into_inner();
-            Box::new(self.grinder.put(
-                BobKey{
-                    key: param.key.as_ref().unwrap().key
-                    },
-                BobData{
-                    data: param.data.take().unwrap().data
-                },
-                {
-                    let mut opts: BobOptions = Default::default();
-                    if let Some(vopts) = param.options.as_ref() { 
-                        if vopts.force_node {
-                            opts |= BobOptions::FORCE_NODE;
+            let sw = Stopwatch::start_new();
+            let param = req.into_inner();
+
+            if !self.put_validate(&param) {
+                warn!("PUT[-] invalid arguments - key and data is mandatory");
+                Box::new(future::err(tower_grpc::Status::new(tower_grpc::Code::InvalidArgument, "Key and data is mandatory")))
+            } else {
+                let key = BobKey{
+                            key: param.key.unwrap().key
+                        };
+                let data = BobData{
+                            data: param.data.unwrap().data
+                        };
+
+                debug!("PUT[{}] data size: {}", key, data.data.len());
+                Box::new(self.grinder.put(key,data,
+                    {
+                        let mut opts: BobOptions = Default::default();
+                        if let Some(vopts) = param.options.as_ref() { 
+                            if vopts.force_node {
+                                opts |= BobOptions::FORCE_NODE;
+                            }
                         }
+                        opts
                     }
-                    opts
-                }
-            ).then(|_r| ok(Response::new(OpStatus {
-                error: None
-            }))))
+                ).then(move |r| {
+                    let elapsed = sw.elapsed_ms();
+                    match r {
+                        Ok(r_ok) => {
+                            info!("PUT[{}]-OK local:{} ok dt: {}ms", key, !r_ok.is_clustered, elapsed);
```

Comment:
This message will be written for every successful request. It is better to write it at the 'debug' level.

Reply:
this is intentional now